### PR TITLE
null, undefined, NaN and -Infinity create holes

### LINKED ISSUE #49: Handle Null values in grid
I have grids with null values. Countour does not like null or undefined in the array of values.  Any solutions?  I can set the nulls to gridMin-1 and then set the threshold but i still Get steps down from the Larger than min values to gridMin.

### PR DESCRIPTION
**Pull request by Fil**
Also, ignore infinite values when computing the thresholds.

Demo: https://observablehq.com/@d3/d3-contours-treats-nulls-as-holes-61

closes #49

### CODE REVIEW

File: src/contours.js [hunk 1]
```diff
@@ -167,18 +172,23 @@ export default function() {
           xt = x | 0,
           yt = y | 0,
           v0,
-          v1 = values[yt * dx + xt];
-      if (x > 0 && x < dx && xt === x) {
-        v0 = values[yt * dx + xt - 1];
+          v1 = low(values[yt * dx + xt]);
+        if (x > 0 && x < dx && xt === x) {
+        v0 = low(values[yt * dx + xt - 1]);
+        if (v0 == null || isNaN(v0) || !isFinite(v0)) v0 = -1e52;
```
Comment (mbostock) [inline src/contours.js:178]: This…

* is redundant with `low(…)`
* treats positive infinity as negative infinity
* uses a magic number when we can handle infinity exactly

The passed-in `value` should be coerced like this (outside this function, probably in `contour` as I mentioned in the other comment):

```js
value = value === null ? NaN : +value;
if (isNaN(value)) throw new Error(`invalid value: ${value}`);
```

Values we read from `values` should be coerced like this:

```js
v0 = v0 === null ? NaN : +v0;
v1 = v1 === null ? NaN : +v1;
```

Given that, maybe something like this?

```js
const a = value - v0;
const b = v1 - v0;
const d = isFinite(a) && isFinite(b) ? a / b : Math.sign(a) / Math.sign(b);
point[0] = isNaN(d) ? x : x + d - 0.5;
```

**Review by mbostock — CHANGES_REQUESTED** [review]
Per above comments. Thanks for taking on this work!

File: src/contours.js [hunk 2]
```diff
@@ -80,25 +80,30 @@ export default function() {
         fragmentByEnd = new Array,
         x, y, t0, t1, t2, t3;
 
+    function above(index) {
+      const x = values[index];
+      return x == null || isNaN(x) ? false : x >= value;
```
[the remark this answers is not in the document]
Comment (Fil) [reply src/contours.js]: The scoped version appears to be slightly faster (if anything).

**Review by mbostock — APPROVED** [review]

**Comment by Fil** [comment]
Thank you! I don't think this needs an addition to README, it's essentially an extension of the "applicability" (or one could say, a bug fix).

**Pull request merged**